lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require 'erb'
+require 'json'
+require 'open3'
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    TRUE_VALUES = [true, 1, '1', 't', 'T', 'true', 'TRUE'].to_set
+    DEPLOY_METADATA_ANNOTATION = 'kubernetes-deploy.shopify.io/metadata'
+
+    def self.inherited(child_class)
+      DiscoverableResource.child_classes.add(child_class)
+    end
+
+    def self.discover(context:, logger:, server_version:)
+      logger.info("Discovering custom resources:")
+      with_retries { discover_groups(context) }
+      if server_version >= Gem::Version.new('1.7.0')
+        kube_client = v1beta1_crd_kubeclient(context)
+        with_retries { discover_crd(logger: logger, client: kube_client) }
+      end
+    end
+
+    def self.discover_groups(context)
+      kinds = discover_kinds(context)
+      kinds.each_pair do |key, val|
+        klass = get_static_class(kind: key)
+        next unless klass
+        klass.const_set(:GROUP, val[:group]) unless klass.constants.include?(:GROUP)
+        klass.const_set(:VERSION, val[:version]) unless klass.constants.include?(:VERSION)
+      end
+    end
+
+    def self.discover_kinds(context)
+      kinds = {}
+
+      # At the top level there is the core group (everything below /api/v1),
+      rest_client = v1_kubeclient(context).create_rest_client
+      raw_json = rest_client['v1'].get(rest_client.headers)
+      resource_list = JSON.parse(raw_json)
+      v1_group_version = { group: 'core', version: 'v1' }
+      resource_list['resources'].map do |res|
+        kind = res['kind']
+        kinds[kind] = v1_group_version
+      end
+
+      # ...and the named groups (at path /apis/$NAME/$VERSION)
+      rest_client = apis_kubeclient(context).create_rest_client
+      raw_json = rest_client.get(rest_client.headers)
+      group_list = JSON.parse(raw_json)
+      group_list = group_list['groups']
+
+      # Map out all detected kinds to their (preferred) group version
+      group_list.each do |group|
+        preferred_version = group['preferredVersion']['groupVersion']
+        all_versions = group['versions'].map { |version| version['groupVersion'] }
+        # Make sure the preferred version gets checked first.
+        all_versions.delete(preferred_version)
+        all_versions.unshift(preferred_version)
+
+        # Grab kinds from all versions
+        all_versions.each do |group_version|
+          raw_response = rest_client[group_version].get(rest_client.headers)
+          json_response = JSON.parse(raw_response)
+          resources = json_response['resources']
+          resources.each do |res|
+            kind = res['kind']
+            next if kinds.key?(kind) # Respect the preferred version
```

Comment:
For each group we'll attempt to detect resources under the preferred group version, but some resources are only available under newer (beta/alpha) versions, so we have to check those too.